Stm web server control

zachmprince opened the pull request:
Closes #31171

Some additional features enabled from this effort:

- batch-restore no longer resets between timesteps
- InputMatrix is now controllable
- StochasticReporters can be executed on `EXEC_TRANSFER` to make sure it has the proper size when the transfer is inserting values and gather on ROOT earlier.
- The python in stochastic tools is reorganized to look like an actual python module.
- Previous STM python was actually never tested, now it is.

zachmprince committed:
Adding StochasticControl #31171

zachmprince committed:
Do not reset apps on each timestep when doing batch-restore #31171

zachmprince committed:
Adding documentation and example of StochasticControl #31171

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3134819/), step Docs: sync website on 16147ed wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31303/site)

This comment will be updated on new commits.

loganharbour commented on modules/stochastic_tools/python/stochastic/StochasticControl.py:
```diff
@@ -0,0 +1,625 @@
+        Clean up after a run by terminating or killing the MOOSE process
+        and removing the temporary input file.
+        """
+        # Terminate gracefully
+        if exc_type is None:
+            # If moose is waiting, send a terminate
+            try:
+                waiting_flag = self.getWaitingFlag()
+            except:
+                waiting_flag = None
+            if not waiting_flag is None:
+                self.setTerminate()
+
+            # Don't finalize if we never started
+            if self._initialized:
+                self.finalize()
```
Can this logic not be somewhat recycled from the MooseControl?

zachmprince replied:
MooseControl will throw if it tries to finalize when the process is waiting. For this application, we want to terminate if the stochastic simulation is finished (the process will still be waiting) and then finalize.

loganharbour commented on modules/stochastic_tools/python/stochastic/StochasticControl.py:
```diff
@@ -0,0 +1,625 @@
+            self._cache.popitem(False)
+
+
+class StochasticRunner:
```
Also missing some docstrings

loganharbour commented on modules/stochastic_tools/test/src/controls/InputMatrixControl.C:
```diff
@@ -0,0 +1,54 @@
+    for (const auto j : make_range(ncol))
+      matrix(i, j) = _sample_function.value(_t, Point(i, j));
+
+  _console << matrix << std::endl;
```
```suggestion
```

loganharbour commented on modules/stochastic_tools/test/src/postprocessors/OptimizationTestFunction.C:
```diff
@@ -0,0 +1,66 @@
+    return eggholder(_x);
+  else
+    paramError("function", "Unknown function: ", _function);
+
+  return 0.0;
```
`paramError` has the `[[ noreturn ]]` modifier so you shouldn't need to return something here

loganharbour commented on modules/stochastic_tools/test/src/postprocessors/OptimizationTestFunction.C:
```diff
@@ -0,0 +1,66 @@
+    return rosen(_x);
+  else if (_function == "eggholder")
+    return eggholder(_x);
+  else
+    paramError("function", "Unknown function: ", _function);
```
Should have this check in the constructor, no?

zachmprince replied:
It's a MooseEnum, show it should error during parsing. I really will never be hit, anyway.

loganharbour commented on modules/stochastic_tools/test/src/postprocessors/OptimizationTestFunction.C:
```diff
@@ -0,0 +1,66 @@
+    return 0.0;
+
+  PostprocessorValue sum = 0.0;
+  for (const auto i : make_range(x.size() - 1))
+    sum += 100.0 * Utility::pow<2>(x[i + 1] - x[i] * x[i]) + Utility::pow<2>(1 - x[i]);
```
No `std::accumulate`? :D 

Also:

```
for (const auto i : index_range(x))
```

instead

zachmprince replied:
Is there a way to `accumulate` when the function considers two neighboring elements? Also, the loop is from 1 to N-1, so I don't think `index_range` can be used unless I break on the last element.

loganharbour reviewed (commented):
Have only looked through the control components

zachmprince committed:
Fixing test failures #31171

- Append stochastic_tools python path using MOOSE_DIR
- Adding issue and design to python tests
- Only skip backup in SamplerFullSolveMultiApp if doing batch-restore

zachmprince committed:
Improving StochasticControl docstrings from review #31171

zachmprince committed:
Addressing rest of initial review #31171

zachmprince committed:
Making sure batch-restore has a backup when initialSetup is re-called #31171

zachmprince committed:
Referencing optimization module version of example #31171

grmnptr commented on modules/stochastic_tools/src/samplers/InputMatrixSampler.C:
```diff
@@ -27,6 +28,20 @@ InputMatrixSampler::InputMatrixSampler(const InputParameters & parameters)
+{
+  if ((dof_id_type)_data.cols() != getNumberOfCols())
+    paramError("matrix",
+               "Inputted matrix as changed the number of columns from ",
```
```suggestion
               "Input matrix changed the number of columns from ",
```

grmnptr commented on modules/stochastic_tools/test/include/postprocessors/OptimizationTestFunction.h:
```diff
@@ -0,0 +1,33 @@
+
+#include "GeneralPostprocessor.h"
+
+class OptimizationTestFunction : public GeneralPostprocessor
```
I know a lot of people are interested in these functions (Dewen was, Som as well, etc). 
I think Som actually tried to add the Rosenbrock function at some point. 
Do you think we could split this function into two different functions and move them to STM from test?
Actually I am fine having it in the test but I still think we should split this in 2 different functions instead of doing the if-else in here.

zachmprince replied:
Made two different function classes with this class being the base.

grmnptr reviewed (changes requested):
Very nice! Almost optimal!

zachmprince committed:
Addressing grammatical errors from review #31171

zachmprince committed:
Separating OptimizationTestFuction into separate objects for each function #31171

zachmprince committed:
Adding option to StochasticControl to keep input when exitting context manager #31171

zachmprince committed:
Making InputMatrix controllable #31171

zachmprince committed:
Attempting to import stochastic tools python for web server tests using executable location #31171

grmnptr reviewed (approved):
Looks good to me! Very cool work.
